lib/google/gax/api_call.rb

```diff
@@ -0,0 +1,136 @@
+# Copyright 2019, Google Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#     * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google Inc. nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+require 'google/gax/api_call/retry_policy'
+require 'google/gax/call_options'
+require 'google/gax/errors'
+require 'time'
+
+module Google
+  module Gax
+    class ApiCall
+      ##
+      # Converts an RPC call into an API call.
+      #
+      # In typical usage, `stub_method` will be a proc used to make an RPC
+      # request. This will mostly likely be a bound method from a request Stub
+      # used to make an RPC call.
+      #
+      # The result is created by applying a series of function decorators
+      # defined in this module to `stub_method`.
+      #
+      # The result is another proc which has the same signature as the original.
+      #
+      # @param stub_method [Proc] used to make a bare rpc call
+      # @param timeout [Numeric] client-side timeout for API calls
+      # @param metadata [Hash]  request headers
+      # @param retry_policy [Hash] the settings for error retry, will be merged
+      #   to the {CallOptions#retry_policy} object if supported.
+      # @param params_extractor [Proc] extracts routing header params from the
```

Comment:
I mentioned this earlier, but I'm still in favor of making this more generic and pushing the routing headers into the generator.

For example, instead of a `params_extractor` that specifically deals with `x-goog-*` parameters could this be a proc that given the request and options returns a hash of headers to be sent?

Reply:
You want the client method to calculate the header off the request object and update the CallOptions object before making the call?

Reply:
If so, can we open a new issue to make this change? I would prefer to do that in a separate PR as it seems orthogonal to retry.

Reply:
More or less. Just saying that there is no reason for this proc to be so specific and I'd like to get the `x-goog-*` stuff out of here.

Separate change is fine - this is a minor thing.

Reply:
Moved to #173.